pkg/controller/ipam/antrea_ipam_controller.go

```diff
@@ -0,0 +1,243 @@
+
+	key := ss.GetUID()
+	c.statefulSetMap.Store(key, ss)
+	c.statefulSetCleanupQueue.Add(key)
+}
+
+// Find IP Pools annotated to StatefulSet via direct annotation or namespace annotation
+func (c *AntreaIPAMController) getIPPoolsByStatefulSet(ss *appsv1.StatefulSet) []string {
+
+	// Inspect pool annotation for the Pods
+	// In order to avoid extra API call in IPAM driver, IPAM annotations are defined
+	// on Pods rather than on StatefulSet
+	selector := map[string]string{"app": ss.Name}
+	podList, err := c.podLister.Pods(ss.Namespace).List(labels.SelectorFromSet(labels.Set(selector)))
+
+	if err != nil {
+		klog.Errorf("Failed to list Pods for StatefulSet %s/%s", ss.Namespace, ss.Name)
+		return nil
+	}
+
+	for _, pod := range podList {
+		annotations, exists := pod.Annotations[AntreaIPAMAnnotationKey]
+		if exists {
+			// Stateful Set Pod is annotated with dedicated IP pool
+			return strings.Split(annotations, AntreaIPAMAnnotationDelimiter)
+		}
+	}
+
+	// Inspect Namespace
+	namespace, err := c.namespaceLister.Get(ss.Namespace)
+	if err != nil {
+		// Should never happen
+		klog.Errorf("Namespace %s not found for StatefulSet %s", ss.Namespace, ss.Name)
+		return nil
+	}
+
+	annotations, exists := namespace.Annotations[AntreaIPAMAnnotationKey]
+	if exists {
+		return strings.Split(annotations, AntreaIPAMAnnotationDelimiter)
+	}
+
+	return nil
+
+}
+
+// Look for an IP Pool associated with this StatefulSet, either a dedicated one or
+// annotated to the namespace. If IPPool is found, this routine will clear all IP
+// addresses that might be reserved for the pool.
+func (c *AntreaIPAMController) cleanIPPoolForStatefulSet(ss *appsv1.StatefulSet) error {
+	klog.InfoS("Processing delete notification", "StatefulSet", ss.Name)
+	poolNames := c.getIPPoolsByStatefulSet(ss)
+	if len(poolNames) == 0 {
+		// no Antrea IPAM annotation for this StatefulSet
+		return nil
+	}
+
+	klog.InfoS("Clearing reserved IP addresses", "StatefulSet", ss.Name, "IP Pools", poolNames)
+	// Only single pool is supported for now.
+	// TODO: Support dual stack.
+	poolName := poolNames[0]
+
+	allocator, err := poolallocator.NewIPPoolAllocator(poolName, c.crdClient, c.ipPoolLister)
+	if err != nil {
+		// This is not a transient error - log and forget
+		klog.Errorf("Failed to find IP Pool %s", poolName)
+		return nil
+	}
+
+	err = allocator.ReleaseStatefulSet(ss.Namespace, ss.Name)
+	if err != nil {
+		// This can be a transient error - worker will retry
+		klog.Errorf("Failed to clean IP allocations for StatefulSet %s in Pool %s", ss.Name, poolName)
+		return err
+
+	}
+
+	return nil
+}
+
+func (c *AntreaIPAMController) statefulSetWorker() {
+	for c.processNextStatefulSetWorkItem() {
+	}
+}
+
+func (c *AntreaIPAMController) processNextStatefulSetWorkItem() bool {
+	key, quit := c.statefulSetCleanupQueue.Get()
+	if quit {
+		return false
+	}
+
+	defer c.statefulSetCleanupQueue.Done(key)
+	ss, ok := c.statefulSetMap.Load(key)
```

Comment:
I did not get why we need a separate map to save StatefulSets, esp. I saw you just use StatefulSet name and Namespace in code.

Reply:
This was from the initial design when statefulset was annotated with IP Pool. The new implementation removed the map.